### ISSUE TITLE
producer#flush doesn't work as expected

### ISSUE THREAD

**Issue by sijie** [apache/pulsar#4642]
**Describe the bug**

The following code example should produce two message batches. but only one message batch is produced.

```
producer.sendAsync("message1");
producer.sendAsync("message2");
producer.flush();

producer.sendAsync("message3");
producer.sendAsync("message4");
producer.flush();
```

**To Reproduce**

Use the example code to reproduce

**Expected behavior**

When `flush` is triggered, it should flush out all the messages.

**Additional context**

The problem is due to following logic. if the connection is not connected, it doesn't send the batch. 

```
private void processOpSendMsg(OpSendMsg op) {
        try {
            batchMessageContainer.clear();
            pendingMessages.put(op);
            ClientCnx cnx = cnx();
            if (isConnected()) {
                // If we do have a connection, the message is sent immediately, otherwise we'll try again once a new
                // connection is established
                op.cmd.retain();
                cnx.ctx().channel().eventLoop().execute(WriteInEventLoopCallback.create(this, cnx, op));
                stats.updateNumMsgsSent(op.numMessagesInBatch, op.batchSizeByte);
            } else {
                if (log.isDebugEnabled()) {
                    log.debug("[{}] [{}] Connection is not ready -- sequenceId {}", topic, producerName,
                        op.sequenceId);
                }
            }
        } catch (InterruptedException ie) {
            Thread.currentThread().interrupt();
            semaphore.release(op.numMessagesInBatch);
            if (op != null) {
                op.callback.sendComplete(new PulsarClientException(ie));
            }
        } catch (Throwable t) {
            semaphore.release(op.numMessagesInBatch);
            log.warn("[{}] [{}] error while closing out batch -- {}", topic, producerName, t);
            if (op != null) {
                op.callback.sendComplete(new PulsarClientException(t));
            }
        }
    }
```

**Comment by jiazhai** [comment]
would like to move to 2.5.0

**Comment by MarvinCai** [comment]
I tried to reproduce and actually get an exception. I did the following
1) Create producer with batch enabled.
2) Produce 3 messages,
3) Set connectionhandler state to Connecting
4) Call flush
5) Set connectionhandler state back to Ready and produce 3 messages then call flush again.
I'm seeing 
`Got ack for msg. expecting: 0 - 2 - got: 3 - 5 - queue-size: 2` 
in log, then connection is closed.
So the first batch is put into pendingMessages queue but not sent out even after connection state is back to Ready.
And when the next batch is sent and ack, first request is pulled from pendingMessages and sequenceId doesn't match.
